File: web/src/app/shared/lib/modelRouting.ts
// Public metadata only. A missing snapshot (older server / failed read) is
// unknown, not an empty routing table. Execution remains the backend's authority.
export type ModelProvider = {
  name: string;
  type: string;
  models: string[];
  catch_all: boolean;
};

export type ModelRouting = ModelProvider[] | null;

// Match the resolver's precedence. Catalog suggestions must not treat a native
// catch-all as an OpenRouter gateway: OpenAI cannot serve a Google catalog slug.
// Explicit routes and listed models remain valid for custom gateways.
export function modelIsAvailable(slug: string, providers: ModelRouting, catalog = false): boolean {
  if (providers === null) return true;
  const value = slug.trim();
  if (!value) return false;
  const slash = value.indexOf("/");
  if (slash > 0 && value.slice(slash + 1).trim() && providers.some((p) => p.name === value.slice(0, slash))) {
    return true;
  }
  if (providers.some((p) => p.models.includes(value))) return true;
  const fallback = providers.find((p) => p.catch_all);
  return !!fallback && (!catalog || fallback.type === "openrouter");
}

export function unavailableModelMessage(slug: string): string {
  return `Model "${slug}" is not available through this workspace's configured providers. Choose a workspace model in the model picker, or ask an admin to configure its provider in Settings → Admin → Model providers.`;
}

File: internal/agentcore/model_routing.go
package agentcore

import (
	"fmt"
	"slices"
	"strings"
)

// ModelProviderInfo is the public routing table, deliberately separate from
// ProviderConfig: credentials and endpoint URLs must never reach the picker.
type ModelProviderInfo struct {
	Name     string       `json:"name"`
	Type     ProviderType `json:"type"`
	Models   []string     `json:"models"`
	CatchAll bool         `json:"catch_all"`
}

// ModelProviders snapshots the ACTIVE resolver, including bundle/env providers
// and successfully applied admin overlays. Reading DB rows instead advertises
// unapplied edits and omits bundle-only deployments.
func (r *ModelResolver) ModelProviders() []ModelProviderInfo {
	out := make([]ModelProviderInfo, 0, len(r.providers))
	for _, p := range r.providers {
		out = append(out, ModelProviderInfo{
			Name: p.Name, Type: p.Type, Models: slices.Clone(p.Models), CatchAll: len(p.Models) == 0,
		})
	}
	return out
}

// CheckModelRoute checks local routing only, without loading a model or calling
// its provider. It uses the same precedence as execution, not a public catalog.
func (r *ModelResolver) CheckModelRoute(slug string) (ProviderType, error) {
	slug = strings.TrimSpace(slug)
	if slug == "" {
		return "", fmt.Errorf("choose a model from the workspace model picker")
	}
	p, _, err := selectProvider(r.providers, slug)
	if err != nil {
		return "", fmt.Errorf("%w; choose a workspace model using <provider-name>/<model>, or ask an admin to configure its provider in Settings → Admin → Model providers", err)
	}
	return p.Type, nil
}

// CatalogModelSlug returns the underlying OpenRouter catalog identifier for a
// routed model. Native/gateway routes keep their identity: stripping a prefix
// from those would borrow another provider's price or context metadata.
func (r *ModelResolver) CatalogModelSlug(slug string) string {
	p, model, err := selectProvider(r.providers, strings.TrimSpace(slug))
	if err == nil && p.Type == ProviderTypeOpenRouter {
		return model
	}
	return slug
}
